Handle `kubelet-arg` when defined in cluster's global configurations

torchiaf opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes https://github.com/rancher/dashboard/issues/10045

### Technical notes summary
<!-- Outline technical changes which may pass unobserved or may help to understand the process of solving the issue -->

Rancher UI expects to find the `kubelet-arg` configuration in `spec.rkeConfig.machineSelectorConfig[0].config` where it can also be defined in `spec.rkeConfig.machineGlobalConfig`, see docs https://ranchermanager.docs.rancher.com/reference-guides/cluster-configuration/rancher-server-configuration/rke2-cluster-configuration#machineselectorconfig.

We are assuming the UI should take the kubelet-arg value from machineGlobalConfig when defined and display it in the same panel where it shows the kubelet-arg from machineSelectorConfig.

Use cases scenarios:

- Create:
    -  only `machineSelectorConfig` is used
- Edit / View:
    - in the case that ONLY machineSelectorConfig in the Yaml, UI can only add/modify kubelet-arg in machineSelectorConfig.
    - in the case that ONLY machineGlobalConfig is defined in the Yaml, UI can only add/modify kubelet-arg in machineGlobalConfig.
    - in the case that BOTH machineSelectorConfig and machineGlobalConfig are defined, the UI handles both and existing args will be updated from where they came from.

### Areas or cases that should be tested
<!-- Areas that should be tested can include Airgap checks, Rancher upgrades, K8s upgrade, etc. -->
<!-- Which browser did you use for local testing? The reviewer should test with a different browser. -->
<!-- Add missing steps or rewrite them if have been missed or to complement existing information. This should define a clear way to reproduce it and not an approximation. -->

- Save the `kubelet-arg` in machineGlobalConfig by manually edit and save cluster's YAML .
- Put the `kubelet-arg` in machineGlobalConfig by Terraform configs, sample file is attached below:

<details><summary>Terraform block</summary>
<p>

```

terraform {
 required_providers {
    rancher2 = {
    source = "rancher/rancher2"
    version = "3.2.0"
    }
  }
}

# Provider bootstrap config
provider "rancher2" {
 api_url = # your rancher url
 insecure = true
 access_key = # your access_key token
 secret_key = # your secret_key token
}

resource "rancher2_machine_config_v2" "rancher2_machine_config_v2" {
 generate_name   = "cfg-terraform"
 digitalocean_config {
  image  = "ubuntu-20-04-x64"
  region  = "nyc1"
  size   = "s-4vcpu-8gb"
 }
}

# Create a new rancher2 RKE2 Cluster
resource "rancher2_cluster_v2" "cfg-terraform" {
 name = "pr-10255-terraform"
 kubernetes_version = "v1.26.10+rke2r2"
 enable_network_policy = false
 default_cluster_role_for_project_members = "admin"
 rke_config {
  # machine_selector_config {
  #   machine_label_selector{
  #     match_labels = {
  #       "key" = "value"
  #     }
  #   }
  # }
  # machine_global_config = <<EOF
  #   kubelet-arg:
  #     - pippo
  # EOF
  machine_pools {
   name = "pool1"
   # this is important
   cloud_credential_secret_name = # DO cluster credentials
   control_plane_role = true
   etcd_role = true
   worker_role = true
   quantity = 1
   machine_config {
    kind = rancher2_machine_config_v2.rancher2_machine_config_v2.kind
    name = rancher2_machine_config_v2.rancher2_machine_config_v2.name
   }
  }
 }
}

```

</p>
</details> 

### Areas which could experience regressions
<!-- Create a detailed list of areas to be analyzed which may be affected by the changes, which would require a prior research to avoid regressions. -->

### Screenshot/Video
<!-- Attach screenshot or video of the changes and eventual comparison if you find it necessary -->

Yaml configs

![image](https://github.com/rancher/dashboard/assets/26394656/daab6079-0466-4f7e-8006-e331ea471282)

UI

![image](https://github.com/rancher/dashboard/assets/26394656/44b61e41-b0fd-4595-9aef-0f26833bcf4a)

richard-cox reviewed (changes requested):
There's a few confusing parts
- How would someone add global level kubelet-args to machineGlobalConfig?
- when viewing yaml in the create rke2 cluster page `machineSelectorConfig` is added at the top level as well as `rkeConfig`. that might be an existing issue
- it looks like we assume the global agent config is the first machineSelectorConfig without a  machineLabelSelector (see shell/models/provisioning.cattle.io.cluster.js `get agentConfig`. should that prefer `spec.rkeConfig.machineGlobalConfig` instead? If so there's a lot of other places the user will report issues
  ```
  get agentConfig() {
    // The one we want is the first one with no selector.
    // If there are multiple with no selector, that will fall under the unsupported message below.
    return this.spec.rkeConfig.machineSelectorConfig.find((x) => !x.machineLabelSelector)?.config;
  }
  ```

I think we might need @nwmac or someone RKE2 side to work through the questions above (and one inline)

torchiaf commented:
Thanks for the review @richard-cox , answers to your points:

1- `kubelet-arg` can be modified at global level using the 'Add global argument' button (see 2nd snapshot) when its value is already defined in the yaml; the UI allows to save the value only in spec.rkeConfig.machineGlobalConfig[0].config, which is how it works on master. Global kubelet-arg can be created only via Yaml, then via terraform. The change was intended to support 'edge' cases , so when cluster settings are defined by Terraform.
We can change this logic though.

2- Could you provide more details ? I went to creation page, added some kubelet-arg, then went to view Yaml and everything looks good:

    ```machineSelectorConfig
      rkeConfig:
        machineSelectorConfig:
          - config:
              kubelet-arg:
                - adwawd
                - awdwd
              protect-kernel-defaults: false
    ```
    
3- Reading the docs, settings have same effects if defined in both `GlobalMachineConfig` and first element in `machineSelectorConfig` array where machineLabelsSelector is empty, https://ranchermanager.docs.rancher.com/reference-guides/cluster-configuration/rancher-server-configuration/rke2-cluster-configuration#machineselectorconfig , so I think we are supporting both cases and preferring settings defined in `agentConfig`.
In general I agree with checking all other places where `agentConfig` is used and put a fallback to `globalMachineConfigs`; I would open a new issue for that.

torchiaf commented on shell/edit/provisioning.cattle.io.cluster/tabs/Advanced.vue:
```diff
@@ -104,11 +107,40 @@ export default {
           </h3>
 
           <ArrayList
+            v-if="i === 0 && serverConfig['kubelet-arg']"
```
I used customer's terraform file, but here is the settings definition https://github.com/rancher/terraform-provider-rancher2/blob/376656c62c6ee854a9f70dd8e2b743b29ccbe100/docs/resources/cluster_v2.md?plain=1#L364 and it seems the relation is spec.rkeConfig.machineGlobalConfig <=> spec.rkeConfig.machineSelectorConfig.config

richard-cox commented:
Ok, this is starting to make sense
- When there's NO spec.rkeConfig.machineGlobalConfig['kubelet-args'] they see below. 
  - ![image](https://github.com/rancher/dashboard/assets/18697775/38698e31-07a7-498b-8426-aa461d65e483)
  - The `For any machines, use the Kubelet args` section will add an entry to spec.rkeConfig.machineSelectorConfig which contains no selector (aka they apply globally)
  - The `For machines with labels matching` section will add an entry to spec.rkeConfig.machineSelectorConfig which contains a selector
- When there IS spec.rkeConfig.machineGlobalConfig['kubelet-args'] they see the UI in your screenshot which allows users to `Add Global Argument` to update `spec.rkeConfig.machineGlobalConfig['kubelet-args'] `

I'm wondering if there's a use case where both global worlds exists (spec.rkeConfig.machineGlobalConfig['kubelet-args'] AND a spec.rkeConfig.machineSelectorConfig entry with no selector). My concern is that the UX is a bit odd, having two lists for `For any machines, use the Kubelet args` and `Add Global Argument`, when to the user they do the same thing. Can we run through this with Ken and Neil after the sync tomorrow?

richard-cox commented:
> 2- Could you provide more details ? I went to creation page, added some kubelet-arg, then went to view Yaml and everything looks good:

This is what i see when going straight to Edit Yaml

![image](https://github.com/rancher/dashboard/assets/18697775/2fb2574f-f312-458e-909b-39cac2f952d9)

I think this might be a previous bug, but want to validate that it won't interfere with these changes

torchiaf commented:
> > 2- Could you provide more details ? I went to creation page, added some kubelet-arg, then went to view Yaml and everything looks good:
> 
> This is what i see when going straight to Edit Yaml
> 
> ![image](https://private-user-images.githubusercontent.com/18697775/301486721-2fb2574f-f312-458e-909b-39cac2f952d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3ODc0NzgsIm5iZiI6MTcwNjc4NzE3OCwicGF0aCI6Ii8xODY5Nzc3NS8zMDE0ODY3MjEtMmZiMjU3NGYtZjMxMi00NThlLTkwOWItMzljYWMyZjk1MmQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAxVDExMzI1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkMjExMWI1ZjE1ZWQzNzJiY2ExMzVlMWMxM2I5ZGYxY2QwZTUzY2FhNmJkZTAyNTBiMWMxZjQ0YTEwYjJjODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.viD0wIstM8-N1Ze5elqzjYlQdETsYXWBHaPB9dDWnD0)
> 
> I think this might be a previous bug, but want to validate that it won't interfere with these changes



> > 2- Could you provide more details ? I went to creation page, added some kubelet-arg, then went to view Yaml and everything looks good:
> 
> This is what i see when going straight to Edit Yaml
> 
 Weird, let me double check.

torchiaf commented:
> I'm wondering if there's a use case where both global worlds exists (spec.rkeConfig.machineGlobalConfig['kubelet-args'] AND a spec.rkeConfig.machineSelectorConfig entry with no selector). My concern is that the UX is a bit odd, having two lists for `For any machines, use the Kubelet args` and `Add Global Argument`, when to the user they do the same thing. Can we run through this with Ken and Neil after the sync tomorrow?

Sure, this is a UX design matter. In my opinion, the cleanest and safer solution would be to add a new section to handle global configs , so there would be a 1:1 match with the YAML, but let's have an overview together.

torchiaf commented:
@richard-cox duplicated `machineSelectorConfig` is also present in `v2.8.0`. I think the one at top level is just ignored by rancher, but needs to be fixed.

richard-cox commented on shell/edit/provisioning.cattle.io.cluster/tabs/Advanced.vue:
```diff
@@ -101,14 +115,48 @@ export default {
+          </ArrayList>
+
+          <ArrayList
+            v-if="row.value.config && (row.value.config['kubelet-arg'] || !serverConfig['kubelet-arg'])"
```
i need to play around with this more, but i think it should be below  i.e. show if there IS a machineSelectorConfig without a selector and with 'kubelet-arg' or there is no machineGlobalConfig a 'kubelet-arg'

```
v-if="row.value.config && ((agentConfig || {})['kubelet-arg'] || !serverConfig['kubelet-arg'])"
```

to test i've been tweaking 

```
  created() {
    this.value.spec.rkeConfig.machineGlobalConfig['kubelet-arg'] = ['imfromglobal'];
    this.value.spec.rkeConfig.machineSelectorConfig[0]['kubelet-arg'] = ['imfromselector'];
  },
```

torchiaf replied:
I changed `row.value.config['kubelet-arg']` -> `(agentConfig || {})['kubelet-arg']`

richard-cox replied:
That change has been reverted, how come?

torchiaf replied:
@richard-cox 

- I had a quick look to the `agentConfig` value and it's the same of row.value.config in this case; it's tested in the unit test though
-  The unit tests fail when using `agentConfig`, since it's a computed value which triggers a DOM update not cacthed in time by unit test.

richard-cox reviewed (commented):
Still reviewing, just updating. 

There's also some strangeness going on with value.spec.rkeConfig.machineSelectorConfig[0]. looks like its populated via the kub3 version's agentArgs which is then overwritten. that's existing functionality, not a regression here

richard-cox reviewed (changes requested):
Just one comment to go https://github.com/rancher/dashboard/pull/10255#discussion_r1483028473

richard-cox reviewed (approved):
I made a small improvement, not related to changes in this PR but in the area.

More detail in the commit message, but tweaked a translation to avoid confusing `For all machines..` --> `For any machines..` text when adding the first selector below
